hasher-matcher-actioner/hmalib/models.py

```diff
@@ -59,51 +67,217 @@ class PipelinePDQHashRecord(PDQRecordBase):
+                    item["SK"], item["SignalSource"]
+                ),
+                item["SignalSource"],
+                item["SignalHash"],
+            )
+            for item in items
+        ]
+
+
+class HashRecordQuery:
+    @staticmethod
+    def from_content_key(
+        table: Table, content_key: str, hash_type_key: str = None
+    ) -> t.List[t.Dict]:
+        """
+        Given a content key (and optional hash type), return its content hash (for that type).
+        Written to be agnostic to hash type so it can be reused by other types of 'HashRecord's.
+        """
+        if hash_type_key is None:
+            key_con_exp = Key("PK").eq(content_key) & Key("SK").begins_with(
+                DynamoDBItem.SIGNAL_KEY_PREFIX
+            )
+        else:
+            key_con_exp = Key("PK").eq(content_key) & Key("SK").eq(hash_type_key)
+
+        return table.query(
+            KeyConditionExpression=key_con_exp,
+            ProjectionExpression="PK, ContentHash, UpdatedAt, Quality",
+        ).get("Items", [])
+
+
+class MatchRecordQuery:
+
+    """
+    Written to be agnostic to hash type so it can be reused by other types of 'MatchRecord's.
+    """
+
+    DEFAULT_PROJ_EXP = "PK, ContentHash, UpdatedAt, SK, SignalSource, SignalHash"
+
+    @classmethod
+    def from_content_key(
+        cls,
+        table: Table,
+        content_key: str,
+        source_prefix: str = DynamoDBItem.SIGNAL_KEY_PREFIX,
+        hash_type: str = None,
+    ) -> t.List[t.Dict]:
+        """
+        Given a content key (and optional hash type), give me its content hash (for that type).
+
+        """
+        filter_exp = None
+        if not hash_type is None:
+            filter_exp = Attr("HashType").eq(hash_type)
+
+        return table.query(
+            KeyConditionExpression=Key("PK").eq(content_key)
+            & Key("SK").begins_with(source_prefix),
+            ProjectionExpression=cls.DEFAULT_PROJ_EXP,
+            FilterExpression=filter_exp,
+        ).get("Items", [])
+
+    @classmethod
+    def from_signal_key(
+        cls,
+        table: Table,
+        signal_key: str,
+        hash_type: str = None,
+    ) -> t.List[t.Dict]:
+        """
+        Given a Signal ID/Key (and optional hash type), give me any content matches found
+        """
+        filter_exp = None
+        if not hash_type is None:
+            filter_exp = Attr("HashType").eq(hash_type)
+
+        return table.query(
+            IndexName="GSI-1",
+            KeyConditionExpression=Key("GSI1-PK").eq(signal_key),
+            ProjectionExpression=cls.DEFAULT_PROJ_EXP,
+            FilterExpression=filter_exp,
+        ).get("Items", [])
+
+    @classmethod
+    def from_time_range(
+        cls, table: Table, hash_type: str, start_time: str = None, end_time: str = None
+    ) -> t.List[t.Dict]:
+        """
+        Given a hash type and time range, give me all the matches found for that type and time range
```

Comment:
Can you be more specific here? Does between include both bounds or only one, or none? Would make a huge difference when we implement pagination.